### ISSUE TITLE
Try to install dali via Pip

### ISSUE THREAD

**Issue by MauritsBleeker** [NVIDIA/DALI#476]
pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali                                 
Looking in indexes: https://pypi.org/simple, https://developer.download.nvidia.com/compute/redist                                                                
Collecting nvidia-dali                                                                                                                                           
  Could not find a version that satisfies the requirement nvidia-dali (from versions: tf-plugin-0.6.1)                                                           
No matching distribution found for nvidia-dali

**Comment by JanuszL** [comment]
Hi,
It is strange. Indeed we have problems with `nvidia-dali-tf-plugin` itself, we are working to solve them, but there is not dependency `nvidia-dali` -> `nvidia-dali-tf-plugin`. So unless you are trying to install `nvidia-dali-tf-plugin` this should work fine.
Could you double check that you cannot install `nvidia-dali` itself?

**Comment by MauritsBleeker** [comment]
Hi,

Thanks for your reply. What do you mean with install nvidia-dali itself? 

pip install nvidia-dali , or install from source?

This also not working.

Thanks,
Maurits

**Comment by cliffwoolley** [comment]
Can you try the original command again now, please?  And, should that still not work, you can try giving the exact version like this as a workaround:
```
pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali==0.6.1
```

**Comment by cliffwoolley** [comment]
We believe this one to be fixed now - please reopen if you find otherwise.  Thanks!

**Comment by MauritsBleeker** [comment]
Hi,

No, it is still not working.

`pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali==0.6.1`

Error message:

Looking in indexes: https://pypi.org/simple, https://developer.download.nvidia.com/compute/redist
Collecting nvidia-dali==0.6.1
Could not find a version that satisfies the requirement nvidia-dali==0.6.1 (from versions: )
No matching distribution found for nvidia-dali==0.6.1

**Comment by MauritsBleeker** [comment]
`pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali`

Is not working as well

**Comment by JanuszL** [comment]
Does the direct link work https://developer.download.nvidia.com/compute/redist/nvidia-dali/nvidia_dali-0.6.1-608405-cp27-cp27mu-manylinux1_x86_64.whl ?
Are you behind any proxy?

**Comment by MauritsBleeker** [comment]
Yes, I can download the .whl file from:  https://developer.download.nvidia.com/compute/redist/nvidia-dali/nvidia_dali-0.6.1-608405-cp27-cp27mu-manylinux1_x86_64.whl

**Comment by JanuszL** [comment]
So it may be something with the page caches. I will ask the person responsible. As I understand you are located in Europe?

**Comment by MauritsBleeker** [comment]
Yes, I'm in Europe. I tried to install nvidia-dali on different machines/clusters, on different networks etc. and python versions to be sure that that is not the problem.

**Comment by JanuszL** [comment]
Please check if that works now.
I'm located in Poland and it works for me (common, cable TV internet provider).

**Comment by rajkaramchedu** [comment]
The http://developer.download.nvidia.com/ itself leads to a 404.

**Comment by JanuszL** [comment]
I'm just using `pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali`.
You can check `https://developer.download.nvidia.com/compute/redist/nvidia-dali`, there you should have list of all binaries.

**Comment by MauritsBleeker** [comment]
MacOS, python 3.7.1, pip 18.1. 
But also tried it on a Linux server centeOS 7. \

Also, for a colleague of mine, it is not working as well. We also tried it on different networks .

**Comment by Kh4L** [comment]
@MauritsBleeker Hi Maurits, as for machine, currently we do not support MacOS

**Comment by JanuszL** [comment]
And Python 3.7. We hope to provide 3.7 version as well in the next release so stay tuned.

**Comment by MauritsBleeker** [comment]
Okay, thanks for your reply! So Python <= 3.6 on the on the Linux server should work? 

Thanks,
Maurits

**Comment by MauritsBleeker** [comment]
okay great. It is working on our Linux server now! Thanks for all the support (and sorry that I never paid much attention to the requirements ...)

**Comment by JanuszL** [comment]
No problem. I'm glad we could help.

**Comment by devillove084** [comment]
When i try to use source code for python 3.7 ...... 

`/usr/bin/ld: /usr/local/lib/libprotobuf.a(arena.o): relocation R_X86_64_TPOFF32 against hidden symbol `_ZN6google8protobuf8internal9ArenaImpl13thread_cache_E' can not be used when making a shared object
/usr/bin/ld: /usr/local/lib/libprotobuf.a(descriptor.o): relocation R_X86_64_PC32 against symbol `_ZTINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [dali/CMakeFiles/dali.dir/build.make:2012: dali/python/nvidia/dali/libdali.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:431: dali/CMakeFiles/dali.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
`

And i use protobuf 3.6.1.

**Comment by JanuszL** [comment]
Hi,
It looks like some protobuf incompatibility, but I'm not 100% sure (maybe some protobuff build option you selected is wrong?). We are using 3.5.1, in [Dockerfile.deps](https://github.com/NVIDIA/DALI/blob/master/Dockerfile.deps) you can see how we build it.
Have you tried [Docker based build](https://github.com/NVIDIA/DALI#compiling-dali-from-source-docker). It may take a while at the first run to the create image with the dependencies but it should produce you exactly the same whl binary as we build.

**Comment by devillove084** [comment]
I juse want build whl binary for python3.7. Tensorflow, Mxnet and PyTorch support python3.7. So, i decide to wait your team. 

By the way, pip install is still not working. And i can not find `nvidia-dali-tf-plugin ` package. How can i download this package.

**Comment by JanuszL** [comment]
> I juse want build whl binary for python3.7. Tensorflow, Mxnet and PyTorch support python3.7. So, i decide to wait your team.

You can do it yourself using `one click build` I have provided.

> By the way, pip install is still not working.

Can you provide more details? As we said before 3.7 is not yet supported by us and we support only Linux platform - no Mac or Windows.

> And i can not find `nvidia-dali-tf-plugin ` package. How can i download this package.

What you get after issuing?
```
pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-tf-plugin
```

**Comment by devillove084** [comment]
Stil `No matching distribution found for `nvidia-dali-tf-plugin`

**Comment by JanuszL** [comment]
If you are using Python 3.7 that this is expected. We will provide Python 3.7 support in the incoming release.

**Comment by devillove084** [comment]
OK.....Then i am trying recomplie protobuf......

**Comment by jph00** [comment]
@JanuszL perhaps whilst this issue isn't fixed in the release version, perhaps it could be left open, or a clear warning about py3.7 added to the top of the readme? (Or maybe if the 0.7 release is nearly done, this is a good reason to push it out ASAP :) )

BTW, it would be great if there was some CHANGES.md file or similar so we could see easily what version is currently release on pip, and what's new.

**Comment by cliffwoolley** [comment]
@jph00 -- https://github.com/NVIDIA/DALI/releases gives a comprehensive list of all changes in each DALI version.  Hope that helps.

**Comment by cliffwoolley** [comment]
@jph00 - also:
> perhaps whilst this issue isn't fixed in the release version, perhaps it could be left open, or a clear warning about py3.7 added to the top of the readme?

This particular issue was closed because the OP was running not only w/ Py3.7 but also on Mac, which just isn't supported at all.  Py3.7 support is tracked in https://github.com/NVIDIA/DALI/issues/444 .

**Comment by Borda** [comment]
Would you consider an option to add `setup.py` to be able to install it via pip from source, similar way as [opencv-python](https://github.com/skvark/opencv-python) does?

**Comment by JanuszL** [comment]
> Would you consider an option to add `setup.py` to be able to install it via pip from source, similar way as [opencv-python](https://github.com/skvark/opencv-python) does?

I would prefer not to do it or let the community maintain it. If you just take a look at the [list of dependencies](https://github.com/NVIDIA/DALI/blob/master/docker/Dockerfile.deps) - (most of them, if not all) have versions way ahead what you can install by default in common Linux distro. So it would require the user to install everything (compile and install) from scratch what may lead to many problems related to some particular configuration the user may have.
Long story short, I see this very hard to do in a way that we can guarantee that it builds at all and that the user gets similar performance to a prebuild version.

**Comment by anubhav4sachan** [comment]
Can easily be solved by installing via `pip install --extra-index-url https://developer.download.nvidia.com/compute/redist nvidia-dali-cuda100` without facing any issues.

For cuda 11.0: replace `nvidia-dali-cuda100` with `nvidia-dali-cuda110`.

**Comment by JanuszL** [comment]
@anubhav4sachan - the original issue was not about python 3.7 but Mac platform. Now DALI supports 3.7 and 3.8 as well.

**Comment by aminakammoun** [comment]
I need help please 
I can't install nvidia dali
(base) C:\Users\EXTRA-INFO>pip install --extra-index-url https://developer.download.nvidia.com/compute/redist/cuda/10.0 nvidia-dali
Looking in indexes: https://pypi.org/simple, https://developer.download.nvidia.com/compute/redist/cuda/10.0
Collecting nvidia-dali
  Using cached https://developer.download.nvidia.com/compute/redist/cuda/10.0/nvidia-dali/nvidia-dali-0.23.0.tar.gz (1.4 kB)
Building wheels for collected packages: nvidia-dali
  Building wheel for nvidia-dali (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'C:\ProgramData\Anaconda3\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"'; __file__='"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-wheel-bqdo1npx'
       cwd: C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\
  Complete output (31 lines):
  running bdist_wheel
  running build
  installing to build\bdist.win-amd64\wheel
  running install
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\setup.py", line 30, in <module>
      setup(name='nvidia-dali',
    File "C:\ProgramData\Anaconda3\lib\site-packages\setuptools\__init__.py", line 165, in setup
      return distutils.core.setup(**attrs)
    File "C:\ProgramData\Anaconda3\lib\distutils\core.py", line 148, in setup
      dist.run_commands()
    File "C:\ProgramData\Anaconda3\lib\distutils\dist.py", line 966, in run_commands
      self.run_command(cmd)
    File "C:\ProgramData\Anaconda3\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\ProgramData\Anaconda3\lib\site-packages\wheel\bdist_wheel.py", line 259, in run
      self.run_command('install')
    File "C:\ProgramData\Anaconda3\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\ProgramData\Anaconda3\lib\distutils\dist.py", line 985, in run_command
      cmd_obj.run()
    File "C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\setup.py", line 22, in run
      raise NameError( \
  NameError:
  ###########################################################################################################
  ERROR: Package nvidia-dali is now deprecated. Please install nvidia-dali-cuda100 instead:
  pip install --extra-index-url https://developer.download.nvidia.com/compute/redist\ nvidia-dali-cuda100
  For more information, go to https://docs.nvidia.com/deeplearning/dali/user-guide/docs/installation.html#installing-prebuilt-dali-packages
  ###########################################################################################################

  ----------------------------------------
  ERROR: Failed building wheel for nvidia-dali
  Running setup.py clean for nvidia-dali
Failed to build nvidia-dali
Installing collected packages: nvidia-dali
    Running setup.py install for nvidia-dali ... error
    ERROR: Command errored out with exit status 1:
     command: 'C:\ProgramData\Anaconda3\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"'; __file__='"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-record-ei9zpmjk\install-record.txt' --single-version-externally-managed --compile --install-headers 'C:\ProgramData\Anaconda3\Include\nvidia-dali'
         cwd: C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\
    Complete output (22 lines):
    running install
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\setup.py", line 30, in <module>
        setup(name='nvidia-dali',
      File "C:\ProgramData\Anaconda3\lib\site-packages\setuptools\__init__.py", line 165, in setup
        return distutils.core.setup(**attrs)
      File "C:\ProgramData\Anaconda3\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "C:\ProgramData\Anaconda3\lib\distutils\dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "C:\ProgramData\Anaconda3\lib\distutils\dist.py", line 985, in run_command
        cmd_obj.run()
      File "C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-install-13nyxd8w\nvidia-dali\setup.py", line 22, in run
        raise NameError( \
    NameError:
    ###########################################################################################################
    ERROR: Package nvidia-dali is now deprecated. Please install nvidia-dali-cuda100 instead:
    pip install --extra-index-url https://developer.download.nvidia.com/compute/redist\ nvidia-dali-cuda100
    For more information, go to https://docs.nvidia.com/deeplearning/dali/user-guide/docs/installation.html#installing-prebuilt-dali-packages
    ###########################################################################################################

    ----------------------------------------
ERROR: Command errored out with exit status 1: 'C:\ProgramData\Anaconda3\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"'; __file__='"'"'C:\\Users\\EXTRA-INFO\\AppData\\Local\\Temp\\pip-install-13nyxd8w\\nvidia-dali\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\EXTRA-INFO\AppData\Local\Temp\pip-record-ei9zpmjk\install-record.txt' --single-version-externally-managed --compile --install-headers 'C:\ProgramData\Anaconda3\Include\nvidia-dali' Check the logs for full command output.

**Comment by JanuszL** [comment]
@aminakammoun - DALI doesn't support the Windows platform so you pip cannot find an appropriate package. Please use a Linux based one to evaluate DALI.

**Comment by jambestwick** [comment]
> [@aminakammoun](https://github.com/aminakammoun) - DALI doesn't support the Windows platform so you pip cannot find an appropriate package. Please use a Linux based one to evaluate DALI.

It's sad to hear this ，I use Windows work computer

**Comment by JanuszL** [comment]
@jambestwick - you can use [WSL](https://learn.microsoft.com/en-us/windows/wsl/install) to run Linux packages on Windows.
